File: hw/total/white_box/BST_remove/code/python/remove.py
from BinNode import BinNode


def remove(BinNode x):
    if not x.has_lc:
        # replace node x with its right child
        return "rChild"
    elif not x.has_rc:
        # replace node x with its left child
        return "lChild"
    else:
        # node x has two childs
        if x.p == None:
            return "x->rChild"
        else:
            w = x.succ()
            u = w.p
            if u == x:
                return "u->rChild"
            else:
                return "u->lChild"

File: hw/total/white_box/BST_remove/code/java/coverage/remove.java
public class remove {
	public String remove(BinNode x) {
		if (!x.has_lc())
			return "rChild";
		else if (!x.has_rc())
			return "lChild";
		else {
			if (x.p == null)
				return "x.rChild";
			else {
				BinNode w = x.succ();
				BinNode u = w.p;
				if (u == x)
					return "u.rChild";
				else
					return "u.lChild";
			}
		}
	}
}
